### PR TITLE
Patterns in let statements and function parameters

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: pil-analyzer/src/evaluator.rs [hunk 1]
```diff
@@ -68,11 +68,20 @@ pub fn evaluate_function_call<'a, T: FieldElement>(
```
Comment (chriseth) [inline pil-analyzer/src/evaluator.rs:68]: The above could actually be an assert now.

File: book/src/pil/patterns.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+> Note that PIL does not check that patterns in a match expression are exhaustive.
+
+## (Ir-)refutability
+
+A pattern is refutable if there is a value of the correct type that the pattern does not match.
+An example is the pattern `7` since it does not match all integers, or the patten `[x, ..]`, because it
+does not match the empty array.
+
+Refutable patterns are fine in match arms, because if the pattern does not match, the evaluator will
+just continue trying the next match arm, but they are disallowed in let statements an ind function
```
Comment (leonardoalt) [inline book/src/pil/patterns.md:52]:
```suggestion
just continue trying the next match arm, but they are disallowed in let statements and in function
```

File: pil-analyzer/src/expression_processor.rs [hunk 3]
```diff
@@ -174,20 +174,17 @@ impl<D: AnalysisDriver> ExpressionProcessor<D> {
...
-            params
-                .iter()
-                .zip(self.local_variable_counter..)
-                .map(|(p, i)| (p.clone(), i)),
-        );
-        self.local_variable_counter += params.len() as u64;
-
+        for param in params {
+            if !param.is_irrefutable() {
+                panic!("Function parameters must be irrefutable, but {param} is refutable.");
```
Comment (leonardoalt) [inline pil-analyzer/src/expression_processor.rs:184]: we should really clean up these panics...

Comment (chriseth) [reply pil-analyzer/src/expression_processor.rs:184]: You mean use Results instead? I think as long as we don't have source references in sub-expressions, we cannot display a nice error message anyway, so it's fine to use panic  for now

**Review by leonardoalt — APPROVED** [review]

File: pil-analyzer/src/expression_processor.rs (hunk 3 above)
Comment (leonardoalt) [reply pil-analyzer/src/expression_processor.rs:184]: yea I know, just meant one day

Comment (leonardoalt) [reply pil-analyzer/src/expression_processor.rs:184]: image

**Pull request merged**